File: ios/Sources/SellwildSDK/SellwildNativeAdView.swift
// SellwildNativeAdView.swift — renders a Prebid native ad into a default
// template and wires impression / click tracking.
//
// Native, unlike banner/outstream, is not auto-rendered by the fork: Prebid
// fetches demand and hands back a `NativeAd` of raw assets (title, body, icon,
// main image, CTA, sponsoredBy). We lay them out here and call
// `nativeAd.registerView(...)` so the fork fires the OMID / impression / click
// trackers against our views.
//
// This view is hosted inside `SellwildAdView` when NATIVE_ENABLED resolves on a
// `.prebidOnly` placement — it reuses the same ad slot (see the render-scope
// decision in SellwildNative.swift). The layout below is a standard template:
//
//   ┌─────────────────────────────────────────┐
//   │ [icon] Title                             │
//   │        Sponsored by …                    │
//   │ ┌───────────────────────────────────┐   │
//   │ │            main media             │   │
//   │ └───────────────────────────────────┘   │
//   │ Body copy …                              │
//   │                        [ Call to action ]│
//   └─────────────────────────────────────────┘

import UIKit
import SellwildPrebidSDK

public final class SellwildNativeAdView: UIView {

    // Forwarded to the hosting SellwildAdView's delegate.
    var onLoaded: (() -> Void)?
    var onImpression: (() -> Void)?
    var onClick: (() -> Void)?
    var onFailed: ((Error) -> Void)?

    private let config: SellwildConfig
    private let zoneId: String
    private let maxHeight: CGFloat

    // Strong reference: the fork's NativeAd must outlive fetchDemand or its
    // trackers/click handling are torn down.
    private(set) var nativeAd: NativeAd?

    let iconView = UIImageView()
    let titleLabel = UILabel()
    let sponsoredLabel = UILabel()
    let mediaView = UIImageView()
    let bodyLabel = UILabel()
    let ctaButton = UIButton(type: .system)

    private var imageTasks: [URLSessionDataTask] = []

    public init(config: SellwildConfig, zoneId: String, maxHeight: CGFloat) {
        self.config = config
        self.zoneId = zoneId
        self.maxHeight = maxHeight
        super.init(frame: .zero)
        buildLayout()
    }

    // sellwild-coverage:exclude-begin(crash-guard) init(coder:) traps by design (storyboards are not supported).
    required init?(coder: NSCoder) { fatalError("Use init(config:zoneId:maxHeight:)") }
    // sellwild-coverage:exclude-end

    deinit {
        for task in imageTasks { task.cancel() }
    }

    // MARK: Load

    /// Request native demand and, on a win, bind + register the assets.
    public func load() {
        // Native may use a dedicated placement id (NATIVE_ZID*); resolve it with
        // the mobile-mirroring precedence, falling back to this slot's zoneId.
        let configId = SellwildNative.resolveConfigId(remoteValues: config.remoteValues, zoneId: zoneId)
        let request = SellwildNative.makeRequest(configId: configId)
        // The winning bid's local cache id is exposed via
        // `bidInfo.targetingKeywords?[PrebidLocalCacheIdKey]` and inflated
        // through `NativeAd.create(cacheId:)`.
        Self.environment.fetchDemand(request) { [weak self] bidInfo in
            self?.demandFetched(bidInfo)
        }
    }

    /// The native auction answered. No bids (no fill) is not a failure
    /// (FAILURES.md 4.3). Any other failed result (network, server, timeout,
    /// bad config id) is `ad.prebid_auction.invalid`, as on the banner path.
    /// A winning bid with no cache id, or one the fork cannot turn into an ad,
    /// is `ad.native_create.invalid`. Either way the host hears `nativeNoFill`,
    /// as before.
    func demandFetched(_ bidInfo: BidInfo) {
        guard bidInfo.resultCode == .prebidDemandFetchSuccess else {
            let result = bidInfo.resultCode
            if SellwildAdPolicy.isAuctionFailure(result.rawValue) {
                SellwildFailures.log(code: .adPrebidAuctionInvalid, component: .native, severity: .warn,
                                     message: "the native auction failed: \(result.name())", zoneId: zoneId)
            } else {
                SellwildLog.debug("[SellwildNativeAdView] no native fill — zone \(zoneId), result \(result.name())")
            }
            onFailed?(SellwildAdError.nativeNoFill)
            return
        }
        guard let cacheId = bidInfo.targetingKeywords?[PrebidLocalCacheIdKey] else {
            reportCreateFailure("a native bid won but carried no local cache id")
            return
        }
        guard let ad = NativeAd.create(cacheId: cacheId) else {
            reportCreateFailure("a native bid won but the native ad could not be created from the cache")
            return
        }
        DispatchQueue.main.async { self.bind(ad) }
    }

    private func reportCreateFailure(_ message: String) {
        SellwildFailures.log(code: .adNativeCreateInvalid, component: .native, message: message, zoneId: zoneId)
        onFailed?(SellwildAdError.nativeNoFill)
    }

    // MARK: Bind

    func bind(_ ad: NativeAd) {
        nativeAd = ad
        ad.delegate = self

        titleLabel.text = ad.title
        bodyLabel.text = ad.text
        sponsoredLabel.text = SellwildAdPolicy.sponsoredText(ad.sponsoredBy)
        ctaButton.setTitle(SellwildAdPolicy.callToActionText(ad.callToAction), for: .normal)

        loadImage(ad.iconUrl, into: iconView)
        loadImage(ad.imageUrl, into: mediaView)

        // Register the whole view for impression tracking; the CTA (and title)
        // are the clickable surfaces.
        ad.registerView(view: self, clickableViews: [ctaButton, titleLabel, mediaView])

        onLoaded?()
    }

    // MARK: Layout (pure UIKit — no fork API)

    private func buildLayout() {
        backgroundColor = .clear
        clipsToBounds = true

        iconView.contentMode = .scaleAspectFit
        iconView.clipsToBounds = true
        iconView.layer.cornerRadius = 6
        iconView.setContentHuggingPriority(.required, for: .horizontal)

        titleLabel.font = .boldSystemFont(ofSize: 15)
        titleLabel.numberOfLines = 2

        sponsoredLabel.font = .systemFont(ofSize: 11)
        sponsoredLabel.textColor = .secondaryLabel

        mediaView.contentMode = .scaleAspectFill
        mediaView.clipsToBounds = true
        mediaView.layer.cornerRadius = 8
        mediaView.backgroundColor = UIColor.secondarySystemBackground
        // The media absorbs the vertical squeeze under the height cap; text and
        // CTA keep their intrinsic size so they're never clipped.
        mediaView.setContentCompressionResistancePriority(.defaultLow, for: .vertical)
        mediaView.setContentHuggingPriority(.defaultLow, for: .vertical)

        bodyLabel.font = .systemFont(ofSize: 13)
        bodyLabel.textColor = .label
        bodyLabel.numberOfLines = 3
        for label in [titleLabel, sponsoredLabel, bodyLabel] {
            label.setContentCompressionResistancePriority(.required, for: .vertical)
        }

        ctaButton.titleLabel?.font = .boldSystemFont(ofSize: 14)
        ctaButton.contentEdgeInsets = UIEdgeInsets(top: 8, left: 16, bottom: 8, right: 16)
        ctaButton.backgroundColor = tintColor
        ctaButton.setTitleColor(.white, for: .normal)
        ctaButton.layer.cornerRadius = 8
        ctaButton.setContentHuggingPriority(.required, for: .horizontal)
        // No addTarget here: the CTA is registered in `registerView(clickableViews:)`,
        // so Prebid's click tracker fires `adWasClicked` → `onClick` on tap. A
        // mirror action here would double-count every CTA click.

        // Header row: icon + (title / sponsored)
        let titleStack = UIStackView(arrangedSubviews: [titleLabel, sponsoredLabel])
        titleStack.axis = .vertical
        titleStack.spacing = 2

        let header = UIStackView(arrangedSubviews: [iconView, titleStack])
        header.axis = .horizontal
        header.spacing = 8
        header.alignment = .center

        // Footer row: body grows, CTA hugs the trailing edge.
        let footer = UIStackView(arrangedSubviews: [bodyLabel, ctaButton])
        footer.axis = .horizontal
        footer.spacing = 8
        footer.alignment = .center

        let root = UIStackView(arrangedSubviews: [header, mediaView, footer])
        root.axis = .vertical
        root.spacing = 8
        root.translatesAutoresizingMaskIntoConstraints = false
        addSubview(root)

        // Hard height cap (the render-side backstop). Let the content be shorter
        // than the slot without a conflict (bottom pin is high, not required),
        // and give the media a min it can break under a tight cap.
        let cap = heightAnchor.constraint(lessThanOrEqualToConstant: maxHeight)
        let bottom = root.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -8)
        bottom.priority = .defaultHigh
        let mediaMin = mediaView.heightAnchor.constraint(greaterThanOrEqualToConstant: 60)
        mediaMin.priority = .defaultHigh

        NSLayoutConstraint.activate([
            root.topAnchor.constraint(equalTo: topAnchor, constant: 8),
            bottom,
            root.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 8),
            root.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -8),
            cap,
            iconView.widthAnchor.constraint(equalToConstant: 40),
            iconView.heightAnchor.constraint(equalToConstant: 40),
            mediaMin,
        ])
    }

    /// Loads a native asset image. A missing asset is fine; a URL that is not
    /// http(s), a failed download or an unusable image is `ad.native_image.network`.
    /// Native asset URLs are bidder-supplied, so file:// is refused and the
    /// payload is capped.
    func loadImage(_ urlString: String?, into imageView: UIImageView) {
        guard let urlString else { return }
        guard let url = SellwildSafeURL.imageURL(urlString) else {
            reportImageFailure(message: "native ad image URL is not http(s)", error: nil)
            return
        }
        let task = Self.environment.imageSession.dataTask(with: url) { [weak self] data, response, error in
            switch SellwildImageLoad.outcome(data: data, response: response, error: error) {
            case .image(let image):
                DispatchQueue.main.async { imageView.image = image }
            case .cancelled:
                break
            case .network(let error):
                self?.reportImageFailure(message: "native ad image failed to download", error: error)
            case .invalid(let problem):
                self?.reportImageFailure(message: "native ad image: \(problem.rawValue)", error: nil)
            }
        }
        imageTasks.append(task)
        task.resume()
    }

    private func reportImageFailure(message: String, error: Error?) {
        SellwildFailures.log(code: .adNativeImageNetwork, component: .native, severity: .warn, error: error,
                             message: message, zoneId: zoneId)
    }
}

// MARK: - Environment

extension SellwildNativeAdView {
    /// Where native demand and asset images come from. Partners always get
    /// `live`; tests replace `SellwildNativeAdView.environment`.
    struct Environment {
        /// Runs the native auction (a Prebid Server request).
        var fetchDemand: (NativeRequest, @escaping (BidInfo) -> Void) -> Void
        /// Downloads asset images.
        var imageSession: URLSession

        static let live = Environment(fetchDemand: fetchLiveDemand, imageSession: .shared)

        // sellwild-coverage:exclude-begin(fetch-demand) NativeRequest.fetchDemand sends the Prebid Server request.
        private static let fetchLiveDemand: (NativeRequest, @escaping (BidInfo) -> Void) -> Void = { request, completion in
            request.fetchDemand(completionBidInfo: completion)
        }
        // sellwild-coverage:exclude-end
    }

    static var environment = Environment.live
}

// MARK: - NativeAdEventDelegate
//
// They drive analytics parity with the banner path; the fork's registerView
// still fires the real trackers.

extension SellwildNativeAdView: NativeAdEventDelegate {

    public func adDidLogImpression(ad: NativeAd) {
        onImpression?()
    }

    public func adWasClicked(ad: NativeAd) {
        onClick?()
    }

    /// The ad's cached bid expired. Its lifecycle, not a failure.
    public func adDidExpire(ad: NativeAd) {
        SellwildLog.debug("[SellwildNativeAdView] native ad expired — zone \(zoneId)")
    }
}

File: android/src/main/kotlin/com/sellwild/sdk/SellwildNativeAdView.kt
// SellwildNativeAdView.kt — renders a Prebid native ad into a default template
// and wires impression / click tracking.
//
// Native, unlike banner/outstream, is not auto-rendered by the fork: Prebid
// fetches demand and hands back a PrebidNativeAd of raw assets (title, body,
// icon, main image, CTA, sponsoredBy). We lay them out here and call
// registerViewList(...) so the fork fires the impression / click trackers
// against our views.
//
// Hosted inside SellwildAdView when NATIVE_ENABLED resolves on a PREBID_ONLY
// placement — it reuses the same ad slot. The layout is a standard template:
// icon + title + sponsoredBy on top, main media in the middle, body + CTA at
// the bottom. Built programmatically to avoid an XML dependency.

package com.sellwild.sdk

import android.content.Context
import android.graphics.BitmapFactory
import android.graphics.Color
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.Gravity
import android.view.View
import android.widget.Button
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import java.net.URL
import com.sellwild.prebid.NativeAdUnit
import com.sellwild.prebid.PrebidNativeAd
import com.sellwild.prebid.PrebidNativeAdEventListener
import com.sellwild.prebid.ResultCode

class SellwildNativeAdView(
    context: Context,
    private val config: SellwildConfig,
    private val zoneId: String,
    private val maxHeightDp: Int,
) : FrameLayout(context) {

    // Forwarded to the hosting SellwildAdView's listener.
    var onLoaded: (() -> Unit)? = null
    var onImpression: (() -> Unit)? = null
    var onClick: (() -> Unit)? = null
    var onFailed: ((String) -> Unit)? = null

    // Strong reference: the fork's native ad must outlive fetchDemand or its
    // trackers / click handling are torn down.
    private var nativeAd: PrebidNativeAd? = null
    private var nativeAdUnit: NativeAdUnit? = null

    private val iconView: ImageView
    private val titleView: TextView
    private val sponsoredView: TextView
    private val mediaView: ImageView
    private val bodyView: TextView
    private val ctaButton: Button

    private val mainHandler = Handler(Looper.getMainLooper())

    init {
        val dp = resources.displayMetrics.density
        fun px(v: Int) = (v * dp).toInt()

        iconView = ImageView(context).apply {
            scaleType = ImageView.ScaleType.FIT_CENTER
        }
        titleView = TextView(context).apply {
            textSize = 15f
            setTypeface(typeface, android.graphics.Typeface.BOLD)
            maxLines = 2
        }
        sponsoredView = TextView(context).apply {
            textSize = 11f
            setTextColor(Color.GRAY)
        }
        mediaView = ImageView(context).apply {
            scaleType = ImageView.ScaleType.CENTER_CROP
            setBackgroundColor(Color.parseColor("#EEEEEE"))
        }
        bodyView = TextView(context).apply {
            textSize = 13f
            maxLines = 3
        }
        ctaButton = Button(context).apply {
            isAllCaps = false
            setOnClickListener { onClick?.invoke() }
        }

        // Header: icon + (title / sponsored)
        val titleStack = LinearLayout(context).apply {
            orientation = LinearLayout.VERTICAL
            addView(titleView)
            addView(sponsoredView)
        }
        val header = LinearLayout(context).apply {
            orientation = LinearLayout.HORIZONTAL
            gravity = Gravity.CENTER_VERTICAL
            addView(iconView, LinearLayout.LayoutParams(px(40), px(40)).apply { rightMargin = px(8) })
            addView(titleStack, LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.WRAP_CONTENT, 1f))
        }

        // Footer: body (grows) + CTA (hugs trailing)
        val footer = LinearLayout(context).apply {
            orientation = LinearLayout.HORIZONTAL
            gravity = Gravity.CENTER_VERTICAL
            addView(bodyView, LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.WRAP_CONTENT, 1f).apply { rightMargin = px(8) })
            addView(ctaButton, LinearLayout.LayoutParams(LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.WRAP_CONTENT))
        }

        // Fixed-height root = the hard cap; the media (height 0 + weight 1)
        // absorbs the leftover space, while header/footer stay WRAP_CONTENT so
        // title / sponsoredBy / body / CTA are never squeezed out.
        val root = LinearLayout(context).apply {
            orientation = LinearLayout.VERTICAL
            setPadding(px(8), px(8), px(8), px(8))
            addView(header, LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT))
            addView(mediaView, LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, 0, 1f).apply { topMargin = px(8); bottomMargin = px(8) })
            addView(footer, LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT))
        }
        addView(root, LayoutParams(LayoutParams.MATCH_PARENT, px(maxHeightDp)))
    }

    // MARK: Load

    /** Request native demand and, on a win, bind + register the assets. */
    fun load() {
        val unit = SellwildNative.makeRequest(zoneId)
        nativeAdUnit = unit
        // Pass a Bundle as the ad object so the fork writes the winning cache
        // id into `BUNDLE_KEY_CACHE_ID` (Util.saveCacheId Bundle path). We
        // then hand that id to `PrebidNativeAd.create(cacheId)`.
        val adBundle = Bundle()
        unit.fetchDemand(adBundle) { resultCode ->
            if (resultCode != ResultCode.SUCCESS) {
                if (config.debug) android.util.Log.d("SellwildNativeAdView", "[native] no fill — zone $zoneId, result $resultCode")
                onFailed?.invoke("Native demand request returned no fill ($resultCode).")
                return@fetchDemand
            }
            val cacheId = adBundle.getString(NativeAdUnit.BUNDLE_KEY_CACHE_ID)
            val ad = cacheId?.let { PrebidNativeAd.create(it) }
            if (ad == null) {
                onFailed?.invoke("Native demand won but no PrebidNativeAd could be created.")
                return@fetchDemand
            }
            mainHandler.post { bind(ad) }
        }
    }

    // MARK: Bind

    private fun bind(ad: PrebidNativeAd) {
        nativeAd = ad

        titleView.text = ad.title
        bodyView.text = ad.description
        sponsoredView.text = ad.sponsoredBy?.takeIf { it.isNotEmpty() }?.let { "Sponsored · $it" } ?: "Sponsored"
        ctaButton.text = ad.callToAction?.takeIf { it.isNotEmpty() } ?: "Learn more"

        loadImage(ad.iconUrl, iconView)
        loadImage(ad.imageUrl, mediaView)

        // Register for impression / click tracking. Shaded fork uses
        // `registerView(container, clickableViews, listener)`.
        ad.registerView(this, listOf(ctaButton, titleView, mediaView), object : PrebidNativeAdEventListener {
            override fun onAdClicked() { onClick?.invoke() }
            override fun onAdImpression() { onImpression?.invoke() }
            override fun onAdExpired() {
                if (config.debug) android.util.Log.d("SellwildNativeAdView", "[native] ad expired — zone $zoneId")
            }
        })

        onLoaded?.invoke()
    }

    // MARK: Image loading (dependency-free)

    private fun loadImage(urlString: String?, target: ImageView) {
        val url = urlString?.takeIf { it.isNotEmpty() } ?: return
        Thread {
            val bmp = runCatching {
                URL(url).openStream().use { BitmapFactory.decodeStream(it) }
            }.getOrNull() ?: return@Thread
            mainHandler.post { target.setImageBitmap(bmp) }
        }.start()
    }

    fun destroy() {
        // Native ad units are one-shot (no auto-refresh to stop); just release.
        nativeAd = null
        nativeAdUnit = null
    }
}
